src/main/scala/PSO/PSOFactory.scala

```diff
@@ -19,29 +21,146 @@ class PSOFactory(config: PSOConfiguration) extends Executor {
+        bestSolution, config.problemType, obj)
+      )
+    }
+    //connect particles based on topology
+    topology match {
+      case ParticleTopology.Ring =>
+        formRing(particles)  
+      case ParticleTopology.FullyConnected =>
+        particles.foreach(p => p ! UpdateNeighbourhood(particles.diff(Set(p))))
+      case ParticleTopology.Grid =>
+        formGrid(particles)
+      case ParticleTopology.Toroidal =>
+        formToroidal(particles)
+      case ParticleTopology.Star =>
+        particles.tail.foreach{p =>
+          p ! NewNeighbour(particles.head)
+          particles.head ! NewNeighbour(p)
+        }
+      case ParticleTopology.BinaryTree =>
+        formBinaryTree(particles.toList)
+    }
+    //termination criteria is # of iterations
+    for(i <- 1 to 1000) {
+      particles.foreach(p => p ! ComputeIteration)
+    }
+    //wait for actors to process all messages and print results
+    Thread.sleep(1000)
+    particles.foreach(p => p ! Report)
 
-    val bestSolution = (initialSolutions.sortWith{case ((_,sol1), (_,sol2)) => obj(sol1) > obj(sol2)}).head._2
+    Future{Result(finalValue = 1)}
+  }
 
-    //Create particles from initial solutions
-    val particles = initialSolutions.to[Set].map{case (id, solution) => 
-      system.actorOf(Props(classOf[Particle], id, List(0.0, 0.0), 0.792, 1.4944, 1.4944, 
-    r.nextDouble, r.nextDouble, solution, bestSolution, obj))}
-      //Using a full-connected topology
-      particles.foreach(p => p ! UpdateNeighbourhood(particles.diff(Set(p))))
-
-      //termination criteria is # of iterations
-      for(i <- 1 to 1000) {
-        particles.foreach(p => p ! CalculateNewPosition)
+  def connectParticlePair(a: ActorRef, b: ActorRef) = {
+    a ! NewNeighbour(b)
+    b ! NewNeighbour(a)
+  }
+  
+  def formRing(particles: Set[ActorRef]) = {
+    connectParticlePair(particles.head, particles.takeRight(1).head)
+    formChain(particles)
+  }
+
+  def formChain(particles: Set[ActorRef]): Unit = {
+    if (particles.size > 1) {
+      connectParticlePair(particles.head, particles.tail.head)
+      formChain(particles.tail)
+    }
+  }
+
+  def buildGrid(particles: Set[ActorRef]): 
+      (Int, List[(Int, Int)], Array[Array[ActorRef]]) = {
+    val w = math.sqrt(particleCount)
+    val width = if (w.isValidInt) {
+      w.toInt
+    } else {
+      throw new RuntimeException(
+        "To use this topology particles must fit in a square grid"
+      )
+    }
+    val cols = (1 to particleCount).map(p => (p - 1) % width)
+    val rows = (1 to particleCount).map(p => (p - 1) / width)
+    val pos = rows.zip(cols)
+    println(pos)
+    val particlePos = particles.zip(pos)
+    val particleGrid = (new Array[Array[ActorRef]](width)).map(_ =>
+      new Array[ActorRef](width)
+    )
+    particlePos.foreach(p => particleGrid(p._2._1)(p._2._2) = p._1)
+    (width, pos.toList, particleGrid)
+  }
+
+  def connectGrid(width: Int, positions: List[(Int,Int)],
```

Comment:
I feel like all the topology stuff should be moved to its own file so that its not taking up all this space here